chainstate/tx-verifier/src/transaction_verifier/mod.rs

```diff
@@ -511,77 +543,74 @@ where
         Ok(())
     }
 
-    fn read_block_undo(
-        &self,
-        tx_source: &TransactionSource,
-    ) -> Result<BlockUndo, ConnectTransactionError> {
-        match self.utxo_block_undo.get(tx_source) {
-            Some(entry) => Ok(entry.undo.clone()),
-            None => match tx_source {
-                TransactionSource::Chain(block_id) => {
-                    let block_undo = self
-                        .storage
-                        .get_undo_data(*block_id)?
-                        .ok_or(ConnectTransactionError::MissingBlockUndo(*block_id))?;
-                    Ok(block_undo)
-                }
-                TransactionSource::Mempool => Err(ConnectTransactionError::MissingMempoolTxsUndo),
-            },
-        }
-    }
-
-    fn fetch_block_undo(
-        &mut self,
-        tx_source: &TransactionSource,
-    ) -> Result<&mut BlockUndo, ConnectTransactionError> {
-        match self.utxo_block_undo.entry(*tx_source) {
-            Entry::Occupied(entry) => Ok(&mut entry.into_mut().undo),
-            Entry::Vacant(entry) => match tx_source {
-                TransactionSource::Chain(block_id) => {
-                    let block_undo = self
-                        .storage
-                        .get_undo_data(*block_id)?
-                        .ok_or(ConnectTransactionError::MissingBlockUndo(*block_id))?;
-                    Ok(&mut entry
-                        .insert(BlockUndoEntry {
-                            undo: block_undo,
-                            is_fresh: false,
-                        })
-                        .undo)
-                }
-                TransactionSource::Mempool => Err(ConnectTransactionError::MissingMempoolTxsUndo),
-            },
-        }
-    }
-
-    fn take_tx_undo(
+    fn connect_pos_accounting_outputs(
         &mut self,
-        tx_source: &TransactionSource,
-        tx_id: &Id<Transaction>,
-    ) -> Result<TxUndo, ConnectTransactionError> {
-        let block_undo = self.fetch_block_undo(tx_source)?;
+        tx_source: TransactionSource,
+        tx: &Transaction,
+    ) -> Result<(), ConnectTransactionError> {
+        let input0 = tx.inputs().get(0).ok_or(ConnectTransactionError::MissingOutputOrSpent)?;
```

Comment:
While taking this out of the map is the right thing to do from an abstraction point of view, keep in mind that it has a different effect that before. Before, it used to do this check only if the output purpose is StakePool. Now it does it all the time.

This may or may not be desired. It may be desired if it's an invariant that inputs can never be empty. If it's not desired, the input0 pulling can be done in a lambda that's called inside. Perhaps we should go with the latter, since we're not restricted by having no inputs in the context of PoS.

I'll push a fix for this.